state/txindex/kv/kv.go

```diff
@@ -68,9 +68,61 @@ func (txi *TxIndex) AddBatch(b *txindex.Batch) error {
 	storeBatch := txi.store.NewBatch()
 	defer storeBatch.Close()
 
+	// temp record holder
+	var duplicateCheck = new(abci.TxResult)
+
+	// keep track of successful txs in this block in order to suppress latter ones being indexed.
+	// note that it deliberately does NOT cache unsuccessful ones
+	var successfulTxsInThisBlock = make(map[string]*abci.TxResult)
```

Comment:
The basic idea is like this:
```golang
if tx.success or not store.Indexed(tx.hash) or not store.Get(tx.hash).success {
  index tx
}
```

And `successfulTxsInThisBlock` is an optimization that in case there are duplicated tx in the current block, and the first one is successful, then we don't need to load the previous result. but in case the duplicated txs are in different blocks, we have to check the previous result.

Do you mean to change the logic to this, so we only need to check existence of previous record without decoding it?
```
if tx.success or not store.Indexed(tx.hash) {
  index tx
}
```

With this, a failed record won't override the previous failed record.

Reply:
Ah, so is the concern also for transactions with the same hash that might have come from _previous_ heights? If that's so, then my suggestion won't work—I thought the logic here was only concerned with duplication within a given height (since the indexing is per-height).

Reply:
yeah, the duplicated txs could be included in different blocks.